crates/matrix-sdk/src/client/mod.rs

```diff
@@ -1284,12 +1293,68 @@ impl Client {
     /// # Arguments
     ///
     /// * `room_id` - The `RoomId` of the room to be joined.
-    pub async fn join_room_by_id(
-        &self,
-        room_id: &RoomId,
-    ) -> HttpResult<join_room_by_id::v3::Response> {
+    pub async fn join_room_by_id(&self, room_id: &RoomId) -> Result<room::Joined> {
         let request = join_room_by_id::v3::Request::new(room_id);
-        self.send(request, None).await
+
+        let (tx, mut rx) =
+            mpsc::unbounded_channel::<(StateEvent<RoomMemberEventContent>, room::Room)>();
+
+        let handle = self
+            .add_event_handler({
+                move |event: SyncStateEvent<RoomMemberEventContent>, room: room::Room| {
+                    let tx = tx.clone();
+
+                    async move {
+                        let full_event = event.into_full_event(room.room_id().to_owned());
+
+                        if full_event.membership() == &MembershipState::Join {
+                            if let Err(e) = tx.send((full_event, room)) {
+                                debug!(
+                                    "Sending event from event_handler failed, \
+                                     receiver already dropped: {}",
+                                    e
+                                );
+                            }
+                        }
+                    }
+                }
+            })
+            .await;
+
+        self.send(request, None).await?;
+
+        debug!("waiting for RoomMemberEvent corresponding to requested join");
+
+        let receive_event_future = timeout(
+            async move {
+                loop {
+                    let (event, room): (StateEvent<RoomMemberEventContent>, room::Room) =
+                        rx.recv().await.expect("receive event and room from event handler");
+
+                    if event.state_key().as_str() == self.user_id().expect("user_id").as_str()
```

Comment:
If you move this check into the event handler, you can switch to a oneshot channel and remove the `loop`, which seems more appropriate here.

Reply:
Good recommendation. Although this is not possible with `join_room_by_id_or_alias`, since there is no `room_id` to check against until the response has been made. Should I get the `room_id` with `resolve_room_alias` and just call `join_room_by_id` afterwards or keep the current structure of sending the event and the room through the channel?

Side note: I don't think I can switch to a `oneshot` channel since `tx` needs to implement `Clone` to pass it into the event handler.

Reply:
I get that you can't use add_room_event_handler here, but why does this suggestion not work for the other method?

Reply:
I'm not sure how to deal with that. Using the same endpoint doesn't seem right to me, I guess having two paths for removing the event handler here wouldn't be too bad.

Reply:
By two paths you mean leave the code as is for this function only?